Check mon connection scores post installation of cluster

s-vipin opened the pull request:
Automation of the issues - BZ-1892173, BZ-2174954, BZ-2151501, BZ-2315595

The steps are updated at - CEPH-83602911

Automation steps:-

1. Deploy a Ceph cluster
2. Validate correct number of mons in connection score, correct number of peers for each mon, each mon present in quorum and valid connection score ( score between 0 - 1 ) 
3. Change election strategy to disallow mode and perform validations mentioned in step(2)
4. Change election strategy to connectivity mode and perform validations mentioned in step(2)
5. Add new host as mon and perform validations mentioned in step(2)
6. Remove mon added in step(5) in the cluster and perform validation mentioned in step(2)
7. Restart monitor service using systemctl and perform validations mentioned in step(2)
8. Stop monitor service using systemctland perform validations mentioned in step(2)
9. Start monitor service using systemctl and validations mentioned in step(2)

Pass log for test case :- http://magna002.ceph.redhat.com/ceph/ceph-qe-logs/vips/logs-vipin-mon-conn-score-18-nov-4-39-pm/

Pass log along with entire suite yet to be updated.

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

s-vipin commented:
@pdhiran I have addressed below review comments provided in previous PR https://github.com/red-hat-storage/cephci/pull/4232 
- Move the method to mon workflows
- Please add doc strings, args, examples of method call, return
- Comment exception for add_mon_service() connection check

I have  added new line at end of the file in my local and pushed it to github, but new lines are not showing up in github.

pdhiran commented on suites/squid/rados/tier-2_rados_test-bugfixes.yaml:
```diff
@@ -204,3 +204,10 @@ tests:
       module: test_v1client.py
```
Please add coverage in al the releases. pacifid, Quincy, Reef & Squid.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
         return True
+
+    def connection_score_checks(
+        self, hosts, rados_obj, num_mons, mon_election_obj
```
You would not need to pass any of the params that are actually being passed in the method. 

All the passed objects are already part of the class.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+            mon_nodes = ceph_cluster.get_nodes(role="mon")
+            mon_nums = len(mon_nodes)
+
+            if not mon_workflow_obj.connection_score_checks(
```
We can just have the method call in example. pre-req steps are not necessary.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+                raise Exception("Connection score checks failed")
+
+        Args:
+            hosts: Cluster object of all monitor host
```
All 4 variables can be obtained. We can remove this.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+            if not rados_obj.check_daemon_exists_on_host(
+                host=host.hostname, daemon_type="mon"
+            ):
+                continue
```
I feel that we can add a log statement, saying "Host X has no mon daemon present".

pdhiran replied:
Add log : "mon not present on host X, but mon label present on the host. Continuing onto the next host for testing" & the log level can be info or debug, since we are not failing the test case upon hitting the condition.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+            )
+
+            if mon_status != 1 or mon_status_desc != "running":
+                continue
```
Same as above. Add a log.error statements, indicating that mon daemon on the host is not running.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+                f"cephadm shell ceph daemon mon.{host.hostname} connection scores dump"
+            )
+            out, err = host.exec_command(sudo=True, cmd=cmd)
+            mon_conn_score = json.loads(out)
```
Maybe add a try - catch block here. 

In case the command execution failed, we will see `JSONDecodeError` here. Let's plan and avoid command execution errors.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+            # Compare number of monitors from connection scores dump and
+            # ceph mon dump commands
+            if len(reports) != num_mons:
+                log.error("Invalid number of mons in connection scores dump")
```
Print more details. like no of mons present in cluster vs what is present in cluster.

Let's add as much details as possible in logging, so that debugging in case of failures is easy.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+                # Compare number of peer monitors in connection scores dump and
+                # ceph mon dump command
+                if len(peer_scores) != len(peer_mons_rank):
+                    log.error("Invalid number of peers in connection scores dump")
```
same as above. let's add more info when failure occurs.

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +282,107 @@ def set_tiebreaker_mon(self, host) -> bool:
+                    ):
+                        log.error("Invalid peer_score value in connection scores dump")
+                        return False
+
```
Let's add info statement, something like "verified connection scores on mon / or on host  X, and all scenarios passed" before returning true

pdhiran commented on tests/rados/test_mon_connection_scores.py:
```diff
@@ -0,0 +1,184 @@
+
+        # Changing strategy to 2. i.e. disallowed mode
+        # and perform connection score validation
+        if not mon_election_obj.set_election_strategy(mode="disallow"):
```
All the workflows need to be performed in all the election strategies.

Let's iterate through the election modes [classic, disallow , connectivity] and perform all the steps in order.

Currently, most of the scenarios would be performed only in connectivity mode.

pdhiran reviewed (commented):
Overall looks good. 

Need to rework on the test workflow, and the class method created.

harshkumarRH commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +284,155 @@ def set_tiebreaker_mon(self, host) -> bool:
+                    # the range of 0 to 1
+                    if (
+                        peer_score_detail["peer_score"] < 0
+                        and peer_score_detail["peer_score"] > 1
```
This needs to be an OR comparison

harshkumarRH commented on suites/pacific/rados/tier-2_rados_test-bugfixes.yaml:
Include the bug tracker IDs in the comments at the top of test suite, please do this for all the test suites

s-vipin commented:
Please review success logs: 

Suite :- http://magna002.ceph.redhat.com/ceph/ceph-qe-logs/vips/logs-squid/  
Test case :- http://magna002.ceph.redhat.com/ceph/ceph-qe-logs/vips/logs-vips-squid/

s-vipin commented:
Resolved all review comments. Success logs after addressing review comments : http://magna002.ceph.redhat.com/ceph/ceph-qe-logs/vips/logs-vips-squid-suite/

pdhiran commented on ceph/rados/monitor_workflows.py:
```diff
@@ -281,3 +283,131 @@ def set_tiebreaker_mon(self, host) -> bool:
+                    f"cephadm shell ceph daemon mon.{host.hostname} connection scores dump"
+                )
+                out, err = host.exec_command(sudo=True, cmd=cmd)
+                mon_conn_score = json.loads(out)
```
required to convert

s-vipin commented:
@pdhiran Success logs after making the discussed changes http://magna002.ceph.redhat.com/ceph/ceph-qe-logs/vips/logs-vips-squid-suite-12-46-pm/ 

https://github.com/red-hat-storage/cephci/blob/36f79ce516ec53201e91047dd7ed7502f6d93dd1/ceph/rados/monitor_workflows.py#L302-L305

pdhiran reviewed (commented):
Looks good. Waiting for other reviewers.

pdhiran commented:
looks good

SrinivasaBharath commented:
LGTM.
Verified the test case - https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83602911

pdhiran reviewed (approved):
looks good
